### ISSUE TITLE
Mesos cluster locked up by jenkins-mesos

### ISSUE THREAD

**Issue by geoffroyjabouley** [jenkinsci/mesos-plugin#183]
This PR is a follow-up on the discussion https://groups.google.com/forum/#!topic/jenkins-mesos/OXd-Ol13iOg

Mesos 0.22.1
Jenkins LTS 1.596.3
Jenkins mesos 0.8.0

**Issue**
A single Jenkins Masters "locking" 100% (in fact, all remaining available) of a mesos cluster resources but not being able to create slaves. When this occurs, we often see 2 jenkins mesos frameworks registered from the same evil Jenkins master (the old one locking the resources, the new one using 0%)

By manually restarting the evil Jenkins master (via <jenkins_url>/restart or via marathon if you use it), all resources are freed and the cluster is working again. No need to restart the whole Mesos cluster and all jenkins masters.

It seems to occurs at a point when almost all resources of the mesos cluster are in use for a certain duration, and the evil Jenkins master is "waiting" for a valid offer.
The mesos driver might crash, leaving a orphaned mesos framework not able to release mesos offers...

Btw, configuring mesos offer timeout option to 10min (for example) does not solve the issue.

This is an annoying issue, as not easy to detect, and destructive because we have to restart the incriminated Jenkins master.
I will try to add more information (logs, screenshots) when it happens again.

**Comment by geoffroyjabouley** [comment]
I think i may have found something. See log below from my Jenkins instance, running with GLOG_v=2. Looks like a problem during slave cleanup process. Then a thread deadlock is detected and last line says that slave cleanup thread is still running (get then this trace forever...)

I have now a rogue Jenkins mesos framework 20150826-103430-320783114-5050-21584-3649 locking all mesos resources in an offered state.

![screenshot-10 195 30 19 5050 2015-11-05 14-28-19](https://cloud.githubusercontent.com/assets/2145297/10969764/9bffcbd6-83ca-11e5-8ff9-53e7d4f05ac0.png)
![screenshot-10 195 30 19 5050 2015-11-05 14-27-42](https://cloud.githubusercontent.com/assets/2145297/10969770/a1c53808-83ca-11e5-9b86-c40f9137cdd4.png)
![screenshot-10 195 30 19 5050 2015-11-05 14-30-25](https://cloud.githubusercontent.com/assets/2145297/10969772/a4778c4a-83ca-11e5-9953-27c48ffedf2c.png)

-------------------------------

Nov 05, 2015 10:18:11 AM org.jenkinsci.plugins.mesos.MesosCleanupThread execute
INFO: Marked **mesos-jenkins-50ad940a-9280-4a86-938b-a80bf8988ba6** for deletion
Nov 05, 2015 10:18:11 AM org.jenkinsci.plugins.mesos.MesosCleanupThread$1 run
INFO: Deleting pending node **mesos-jenkins-50ad940a-9280-4a86-938b-a80bf8988ba6**
Nov 05, 2015 10:18:11 AM org.jenkinsci.plugins.mesos.MesosSlave terminate
INFO: Terminating slave **mesos-jenkins-50ad940a-9280-4a86-938b-a80bf8988ba6**
I1105 09:18:11.526890  3478 process.cpp:2160] Resuming scheduler-a3805ab7-c9f3-45cc-a0ed-74999b49963f@10.195.96.117:20002 at 2015-11-05 09:18:11.526832128+00:00
I1105 09:18:11.565681  3478 process.cpp:2160] Resuming scheduler-a3805ab7-c9f3-45cc-a0ed-74999b49963f@10.195.96.117:20002 at 2015-11-05 09:18:11.565661952+00:00
Nov 05, 2015 10:18:11 AM org.jenkinsci.remoting.nio.NioChannelHub run
SEVERE: Unexpected key cancellation for org.jenkinsci.remoting.nio.NioChannelHub$MonoNioTransport@e3de6b2[name=**mesos-jenkins-50ad940a-9280-4a86-938b-a80bf8988ba6**]
java.nio.channels.CancelledKeyException
	at sun.nio.ch.SelectionKeyImpl.ensureValid(SelectionKeyImpl.java:73)
	at sun.nio.ch.SelectionKeyImpl.readyOps(SelectionKeyImpl.java:87)
	at java.nio.channels.SelectionKey.isReadable(SelectionKey.java:289)
	at org.jenkinsci.remoting.nio.NioChannelHub.run(NioChannelHub.java:563)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)

I1105 09:18:11.614933  3480 pid.cpp:82] Attempting to parse 'master@10.195.30.19:5050' into a PID
I1105 09:18:11.615000  3480 process.cpp:577] Parsed message name 'mesos.internal.StatusUpdateMessage' for scheduler-a3805ab7-c9f3-45cc-a0ed-74999b49963f@10.195.96.117:20002 from master@10.195.30.19:5050
I1105 09:18:11.615619  3479 process.cpp:2160] Resuming scheduler-a3805ab7-c9f3-45cc-a0ed-74999b49963f@10.195.96.117:20002 at 2015-11-05 09:18:11.615565056+00:00
I1105 09:18:11.616402  3479 pid.cpp:82] Attempting to parse 'slave(1)@10.195.30.20:5051' into a PID
I1105 09:18:11.616439  3479 sched.cpp:680] Received status update TASK_KILLED (UUID: bc060f85-0275-49ea-a063-5d4ee995ab0d) for task **mesos-jenkins-50ad940a-9280-4a86-938b-a80bf8988ba6** of framework 20150826-103430-320783114-5050-21584-3649 from slave(1)@10.195.30.20:5051
Nov 05, 2015 10:18:11 AM jenkins.slaves.JnlpSlaveAgentProtocol$Handler$1 onClosed
WARNING: NioChannelHub keys=3 gen=4941956: Computer.threadPoolForRemoting [#4] for + **mesos-jenkins-50ad940a-9280-4a86-938b-a80bf8988ba6** terminated
java.io.IOException: Connection aborted: org.jenkinsci.remoting.nio.NioChannelHub$MonoNioTransport@e3de6b2[name=**mesos-jenkins-50ad940a-9280-4a86-938b-a80bf8988ba6**]
	at org.jenkinsci.remoting.nio.NioChannelHub$NioTransport.abort(NioChannelHub.java:211)
	at org.jenkinsci.remoting.nio.NioChannelHub.run(NioChannelHub.java:640)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.nio.channels.CancelledKeyException
	at sun.nio.ch.SelectionKeyImpl.ensureValid(SelectionKeyImpl.java:73)
	at sun.nio.ch.SelectionKeyImpl.readyOps(SelectionKeyImpl.java:87)
	at java.nio.channels.SelectionKey.isReadable(SelectionKey.java:289)
	at org.jenkinsci.remoting.nio.NioChannelHub.run(NioChannelHub.java:563)
	... 6 more

Nov 05, 2015 10:18:55 AM jenkins.metrics.api.Metrics$HealthChecker execute
**WARNING: Some health checks are reporting as unhealthy: [thread-deadlock]**
Nov 05, 2015 10:19:11 AM hudson.model.AsyncPeriodicWork doRun
**INFO: Mesos pending deletion slave cleanup thread is still running. Execution aborted.**

**Comment by gaga-github** [comment]
I have a jenkins master that always requests all the resources even after a restart of the jenkins master.

I've tried uninstalling the mesos plugin and removing the mesos directory under plugins.

Moved it to another mesos cluster, restarted all the masters in a cluster

Same result each time.

**Comment by etiennebec** [comment]
We recently encounter this situation and the explanation is straightforward after analyzing the Jenkins thread dump:

```
Computer.threadPoolForRemoting [#4703]

"Computer.threadPoolForRemoting [#4703]" Id=84698 Group=main WAITING on java.util.concurrent.locks.ReentrantLock$NonfairSync@5805f1a owned by "Thread-36145" Id=84703
    at sun.misc.Unsafe.park(Native Method)
    -  waiting on java.util.concurrent.locks.ReentrantLock$NonfairSync@5805f1a
    at java.util.concurrent.locks.LockSupport.park(LockSupport.java:186)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:834)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(AbstractQueuedSynchronizer.java:867)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:1197)
    at java.util.concurrent.locks.ReentrantLock$NonfairSync.lock(ReentrantLock.java:214)
    at java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:290)
    at org.jenkinsci.plugins.mesos.JenkinsScheduler.supervise(JenkinsScheduler.java:840)
    at org.jenkinsci.plugins.mesos.JenkinsScheduler.terminateJenkinsSlave(JenkinsScheduler.java:240)
    -  locked org.jenkinsci.plugins.mesos.JenkinsScheduler@2d235b9d

    at org.jenkinsci.plugins.mesos.MesosImpl.stopJenkinsSlave(MesosImpl.java:36)
    -  locked org.jenkinsci.plugins.mesos.MesosImpl@7885838f

    at org.jenkinsci.plugins.mesos.MesosComputerLauncher.terminate(MesosComputerLauncher.java:122)
    at org.jenkinsci.plugins.mesos.MesosSlave.terminate(MesosSlave.java:92)
    at org.jenkinsci.plugins.mesos.MesosCleanupThread$1.run(MesosCleanupThread.java:85)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
    at java.util.concurrent.FutureTask.run(FutureTask.java:262)
    at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:745)

    Number of locked synchronizers = 1
    - java.util.concurrent.ThreadPoolExecutor$Worker@267b7fb6

Thread-36145

"Thread-36145" Id=84703 Group=main BLOCKED on org.jenkinsci.plugins.mesos.MesosImpl@7885838f owned by "Computer.threadPoolForRemoting [#4703]" Id=84698
    at org.jenkinsci.plugins.mesos.MesosImpl.stopScheduler(MesosImpl.java:20)
    -  blocked on org.jenkinsci.plugins.mesos.MesosImpl@7885838f
    at org.jenkinsci.plugins.mesos.JenkinsScheduler.supervise(JenkinsScheduler.java:866)
    at org.jenkinsci.plugins.mesos.JenkinsScheduler.statusUpdate(JenkinsScheduler.java:761)

    Number of locked synchronizers = 1
    - java.util.concurrent.locks.ReentrantLock$NonfairSync@5805f1a
```

There's a race condition between the periodic work ``MesosCleanupThread:execute`` **task1** and the Mesos driver callback method ``JenkinsScheduler:statusUpdate`` **task2** (which is invoked "when the status of a task has changed (e.g., a slave is lost and so the task is lost, a task finishes and an executor sends a status update saying so, etc)" according to the Javadoc). It should only happen when the Mesos Cloud is set on demand registration mode. If you follow the trace you will see that the deadlock is caused by:
- **task1** locking ``MesosImpl`` then ``JenkinsScheduler`` whose methods ``stopJenkinsSlave/terminateJenkinsSlave`` are synchronized
- **task2** manually locking ``JenkinsScheduler:supervise`` via a ``ReentrantLock SUPERVISOR_LOCK``
- **task1** waiting for the reentrantlock to be released
- **task2** waiting to run ``MesosImpl:stopScheduler``

One quick workaround we currently use is calling ``SUPERVISOR_LOCK.tryLock()`` in ``JenkinsScheduler:supervise`` instead of ``SUPERVISOR_LOCK.lock()``, which will block the thread until the lock is released. It allows us to avoid the deadlock yet to log the issue. However if it happens after all the builds ran, you might see idling Mesos slave and the Jenkins framework will still be registered as we won't disconnect until ``supervise()`` is called again.

I will do a PR if I find a more suitable fix after digging a little bit more in the code of the plugin.

**Comment by eThomasHerrlin** [comment]
We have seen this problem as well, and we are using Jenkins 1.625.3, Jenkins mesos-0.9.0 plugin, Mesos 0.26.0.

I have a potential fix for orphaned lock. Uncertain if this is the root cause for the problem we are seeing here.

ConcurrentModificationException could in theory happen in the ```JenkinsScheduler.supervise() { ... Mesos.getAllClouds() ... }``` call.
It is outside of the ```try {...} finally {unlock}``` protection.
Suggest moving it inside the try{} block aswell.

getAllClouds() does not seem to be protected from modification by getInstance().
jenkinsci/plugins/mesos/Mesos.java static clouds map could be switched to ConcurrentHashMap to fix potential ConcurrentModificationException issues with getInstance() and getAllClouds() running simultaniously.

**Comment by etiennebec** [comment]
I don't think it is directly related to this issue. If a ConcurrentModificationException was to be thrown it would just crash the plugin and log a stack trace in the Jenkins log (which is not the case).

Furthermore Mesos.getInstance() is a synchronized method which means to my understanding that it will lock the Mesos/MesosImpl object during its execution thus ensuring a consistent state of the hashmap.

**Comment by eThomasHerrlin** [comment]
Still seeing deadlock problems intermittently with the tryLock patch and my own move of getAllClouds into the try{}.
Did not see the "Failed to acquire lock: postpone..." in the logs.

I will now try to cherry-pick #101 aa7ba53 "Fix a deadlock case", as it affects stopJenkinsSlave/stopScheduler locking that we (me and etiennebec) see in our stack-traces.

**Comment by Vlatombe** [comment]
I invite you to give a try with 0.10.0-beta-4 available on the [experimental update center](https://jenkins-ci.org/blog/2013/09/23/experimental-plugins-update-center/). I fixed a couple cases of deadlocks and race conditions that could lead to ghost frameworks, and spurious offer retention.

**Comment by geoffroyjabouley** [comment]
Thanks Vincent we will try this new version soon (when our new mesos
cluster is up and running)

2016-02-01 11:43 GMT+01:00 Vincent Latombe <notifications@github.com>:

**Comment by imrangit** [comment]
I am also seeing bunch of these messages in mesos master logs but for a differnet framework (spark).
W0329 18:30:31.174345  9687 master.cpp:6194] Possibly orphaned completed task 24614 of framework 12175bd1-94e9-487e-b95f-4a916ef54de2-0043 that ran on slave bfe99321-cb8d-4d56-b20c-844d2571298a-S1 at slave(1)@10.3.161.102:5051

Any tuning that we can do on master or slaves to resolve this?

**Comment by gaga-github** [comment]
I see this also on 0.12.0, restarting the Jenkins master doesn't resolve it.  I see it ask for 3 offers (one for each slave).
